### PR TITLE
core: do not leak mount for credentials directory if possible

### PR DESCRIPTION
**Pull request by yuwata**
Note, both open_tree() and move_mount() are since kernel v5.2.
https://github.com/torvalds/linux/commit/a07b20004793d8926f78d63eb5980559f7813404
https://github.com/torvalds/linux/commit/2db154b3ea8e14b04fee23e3fdfd5e9d17fbc6ae

### CODE REVIEW

File: src/core/execute.c [hunk 1]
```diff
@@ -2047,7 +2047,9 @@ static int build_environment(
                 our_env[n_env++] = x;
         }
 
-        if (exec_context_has_credentials(c) && p->prefix[EXEC_DIRECTORY_RUNTIME]) {
+        if (exec_context_has_credentials(c) &&
+            p->prefix[EXEC_DIRECTORY_RUNTIME] &&
+            FLAGS_SET(p->flags, EXEC_WRITE_CREDENTIALS)) {
```
Comment (bluca) [inline src/core/execute.c]: maybe swap the flag check and the function call? saves one lazy evaluation call

File: src/core/credential.c [hunk 2]
```diff
@@ -848,6 +851,23 @@ static int setup_credentials_internal(
                         if (r < 0)
                                 return r;
 
+                        if (ret_mount_fd) {
+                                _cleanup_close_ int mount_fd = -EBADF;
+
+                                r = mount_fd = RET_NERRNO(open_tree(AT_FDCWD, workspace, OPEN_TREE_CLONE | OPEN_TREE_CLOEXEC));
+                                if (r >= 0) {
+                                        r = umount_verbose(LOG_DEBUG, workspace, MNT_DETACH|UMOUNT_NOFOLLOW);
```
Comment (bluca) [inline src/core/credential.c:862]: why the umount?

File: src/core/execute.c (hunk 1 above)
Comment (bluca) [reply src/core/execute.c]: nvm it's removed in a later commit

**Review by bluca — APPROVED** [review]

File: src/core/credential.c (hunk 2 above)
Comment (bluca) [reply src/core/credential.c:862]: please add a short comment to explain it as it's not immediately obvious

**Comment by yuwata** [comment]
Rebased, and added several comments in the code. Upgrading the green label.

**Comment by keszybz** [comment]
I'm not sure if I understand all the consequence of this, but @bluca liked it, and the implementation looks good.

File: src/core/credential.h [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+DEFINE_TRIVIAL_CLEANUP_FUNC(ExecLoadCredential*, exec_load_credential_free);
+
+extern const struct hash_ops exec_set_credential_hash_ops;
+extern const struct hash_ops exec_load_credential_hash_ops;
+
+bool exec_context_has_encrypted_credentials(ExecContext *c);
+bool exec_context_has_credentials(const ExecContext *c);
+
+int exec_context_destroy_credentials(const ExecContext *c, const char *runtime_root, const char *unit);
+int setup_credentials(
```
Comment (poettering) [inline src/core/credential.h:43]: hm, if this is now exported, this should really have an `exec_` prefix too.

File: src/core/credential.h [hunk 4]
```diff
@@ -0,0 +1,45 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
```
Comment (poettering) [inline src/core/credential.h:1]: I think this .c/.h file should really be called `exec-credential.[ch]`, given it's still pretty closely related to execute.[ch], i.e. uses the Exec/exec_ prefix for most its definitions.

**Comment by poettering** [comment]
This is quite a change in behaviour. My original implementation of the credentials logic made sure that the admin can always see what services acquired as credential, and that the credential fs is always shared between ExecStart=/ExecStartPre= invocations, and is even still available if you run ExectStop= (we won't set it up for that, but we'll still set $CREDENTIALS_DIRECTORY to make it accessible if it was already). 

I am quite frankly not sure that this new behaviour is systematically better. And it just leaves the actual issue entirely unaddressed, mkosi boots are still full of failed umounts...

**Comment by yuwata** [comment]
Ah, I did not know that `ExecStop=` has credentials directory. I will fix that. Thank you for the comments.

**Pull request merged**